File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int start = 0;
        int end = 0;
        
        for (int weight : weights){
            start = Math.max(start, weight);
            end += weight;
        }
        
        while(start<=end){
            int mid = start + (end-start)/2;
            int reqDays = 1;
            int tempSum = 0;
            for(int weight : weights){
                tempSum = tempSum + weight;
                if(tempSum>mid){
                    tempSum = weight;
                    reqDays+=1;
                }
            }
            
            if(reqDays <= days){
                end = mid-1;
            }else{
                start = mid+1;
            }
        }
        
        return start;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low = max(weights)
        high = sum(weights)
        result = low
        while(low<=high):
            mid = low + (high - low)//2
            mid_req_days = 0
            temp = 0
            for weight in weights:
                temp += weight
                if temp>mid:
                    temp = 0 
                    temp+=weight
                    mid_req_days+=1
            mid_req_days+=1            
            if mid_req_days <= days:
                high = mid-1
                result = mid
            else:
                low = mid+1
                
        return result